File: apps/RelayiumKit/Sources/RelayiumKit/Realtime/RelayRenewSDP.swift
import Foundation

/// What a renewal is allowed to change about a live PeerConnection, and what it
/// is not.
///
/// A migration restarts ICE on the SAME connection: a new ufrag, new
/// candidates, a new selected pair. It must not be able to change the DTLS peer,
/// the media shape, or the answering side's role — those are what
/// `RelayRenewSDPPin` holds fixed for every description at `epoch >= 1`.
///
/// **Pinning keeps the existing DTLS peer in place. It is NOT the root of
/// trust** — that remains the E2E key the SAS anchored. What it removes is the
/// one thing an authenticated renewal could otherwise be used for: handing a
/// live connection a description that re-points it at a different endpoint.

// MARK: - the pin

/// The three properties of an applied description a renewal may not move.
public struct RelayRenewSDPPin: Equatable, Sendable {
    /// `a=fingerprint` values, normalised to `<hash-lower> <HEX-UPPER>`, sorted
    /// and deduplicated.
    ///
    /// A SET rather than a list, and normalised rather than compared raw: the
    /// two ends may legitimately list the same fingerprints in a different
    /// order, and RFC 8122 leaves the hex case open. A raw comparison would
    /// reject a peer that merely re-rendered its own certificate.
    public let fingerprints: [String]
    /// The `a=mid` sequence, in order.
    public let mids: [String]
    /// How many `m=` lines the description has. Carried alongside `mids`
    /// because an m-line with no mid still changes the media shape.
    public let mLineCount: Int
    /// The first `a=setup` role, or nil. Compared ONLY for an answer: an offer
    /// legitimately restates `actpass`, so the role is only checked where it is
    /// actually chosen.
    public let setup: String?

    public init(fingerprints: [String], mids: [String], mLineCount: Int, setup: String?) {
        self.fingerprints = fingerprints
        self.mids = mids
        self.mLineCount = mLineCount
        self.setup = setup
    }

    /// Whether a description at `epoch >= 1` may be applied against this
    /// baseline.
    ///
    /// All three conditions, and the answer's role on top. Deliberately written
    /// as separate guards rather than one boolean chain so a future fourth
    /// condition cannot be added in a position where `&&` short-circuits past
    /// it.
    public func admits(_ other: RelayRenewSDPPin, as type: RelayRenewSDPType) -> Bool {
        guard fingerprints == other.fingerprints else { return false }
        guard mLineCount == other.mLineCount else { return false }
        guard mids == other.mids else { return false }
        guard type == .answer else { return true }
        return setup == other.setup
    }
}

/// Read the pin out of one SDP.
///
/// Total and lenient about everything it does not name: an SDP carries dozens
/// of attributes a renewal has no opinion about, and the only ones that matter
/// here are the three this extracts. A description with no fingerprint produces
/// an empty set, which can only ever equal another empty set — so a baseline
/// that never carried one cannot be used to admit a description that does.
public func relayRenewPin(sdp: String) -> RelayRenewSDPPin {
    var fingerprints: Set<String> = []
    var mids: [String] = []
    var mLineCount = 0
    var setup: String?

    for rawLine in sdp.split(whereSeparator: { $0 == "\r\n" || $0 == "\n" || $0 == "\r" }) {
        let line = String(rawLine)
        if line.hasPrefix("m=") {
            mLineCount += 1
        } else if let value = sdpAttribute(line, "fingerprint") {
            if let normalised = normalisedFingerprint(value) { fingerprints.insert(normalised) }
        } else if let value = sdpAttribute(line, "mid") {
            mids.append(value)
        } else if let value = sdpAttribute(line, "setup"), setup == nil {
            setup = value
        }
    }
    return RelayRenewSDPPin(fingerprints: fingerprints.sorted(),
                            mids: mids,
                            mLineCount: mLineCount,
                            setup: setup)
}

/// `<hash-lower> <HEX-UPPER>`, or nil for a value that is not a fingerprint.
///
/// The two halves are normalised in OPPOSITE directions on purpose, and that is
/// what the fixture pins: the hash name is a registered token compared
/// case-insensitively, while the hex digits are conventionally upper case and
/// legitimately arrive either way.
private func normalisedFingerprint(_ value: String) -> String? {
    let parts = value.split(separator: " ", maxSplits: 1, omittingEmptySubsequences: true)
    guard parts.count == 2 else { return nil }
    return parts[0].lowercased() + " " + parts[1].uppercased()
}

/// The value of `a=<name>:<value>` on one SDP line, or nil.
private func sdpAttribute(_ line: String, _ name: String) -> String? {
    let prefix = "a=" + name + ":"
    guard line.hasPrefix(prefix) else { return nil }
    return String(line.dropFirst(prefix.count))
}

// MARK: - ufrag binding

/// The `a=ice-ufrag` of one description, or the empty string.
///
/// This is where a LOCAL generation's name comes from: the ufrag of the
/// description whose `setLocalDescription` succeeded for that epoch. Nothing
/// here reads a "current epoch" variable, which is the whole point — see
/// `relayRenewCandidateUfrag`.
public func relayRenewICEUfrag(sdp: String) -> String {
    for rawLine in sdp.split(whereSeparator: { $0 == "\r\n" || $0 == "\n" || $0 == "\r" }) {
        if let value = sdpAttribute(String(rawLine), "ice-ufrag"), !value.isEmpty {
            return value
        }
    }
    return ""
}

/// The generation a candidate NAMES, parsed from its own ` ufrag <x>`
/// extension. The empty string when it names none.
///
/// A candidate belongs to whichever ICE generation the candidate itself names,
/// never to whatever epoch happens to be current when the callback fires.
/// `didGenerate` is asynchronous and a restart can land between gathering and
/// delivery, so labelling by a mutable "current epoch" variable attributes
/// candidates to the wrong generation — which is exactly how an old relay
/// candidate would be presented as proof of a new path.
///
/// Apple's `RTCIceCandidate` exposes no `usernameFragment` property at all, so
/// this parse is the only source on this platform, and a local candidate whose
/// ufrag cannot be parsed is dropped rather than sent.
///
/// The scan starts at the first extension pair rather than at token 0, so a
/// transport, an address or a candidate type that happened to read `ufrag`
/// cannot be mistaken for the extension's name.
public func relayRenewCandidateUfrag(candidate: String) -> String {
    var line = candidate
    if line.hasPrefix("a=") { line = String(line.dropFirst(2)) }
    let tokens = line.split(separator: " ", omittingEmptySubsequences: true).map(String.init)
    // foundation, component, transport, priority, address, port, "typ", type.
    let firstExtension = 8
    guard tokens.count > firstExtension else { return "" }
    var i = firstExtension
    while i + 1 < tokens.count {
        if tokens[i] == "ufrag" { return tokens[i + 1] }
        i += 2
    }
    return ""
}

/// The generation an INBOUND candidate names, reconciling its two possible
/// sources. The empty string means the candidate cannot be bound and must be
/// dropped.
///
/// - Both present and equal, or only one present: that value.
/// - Both present and contradictory: dropped. A peer that names two
///   generations for one candidate has named none this side may act on, and
///   preferring either source would let the disagreement decide which.
/// - Neither present: dropped, for the reason spec §3.2 gives — a candidate
///   whose generation cannot be named cannot be bound to an epoch.
public func relayRenewInboundCandidateUfrag(candidate: String,
                                            usernameFragment: String) -> String {
    let fromString = relayRenewCandidateUfrag(candidate: candidate)
    let fromField = usernameFragment
    if fromString.isEmpty { return fromField }
    if fromField.isEmpty { return fromString }
    return fromString == fromField ? fromString : ""
}

File: apps/android/protocol/src/main/kotlin/com/relayium/protocol/RelayRenewSdp.kt
package com.relayium.protocol

/**
 * SDP pinning and ICE-generation binding — `relay-renew-v1.md` section 5.
 *
 * Pure string work, deliberately: `org.webrtc` types appear nowhere in this
 * module, so every rule here is a plain JVM test against the shared fixture
 * rather than something only an emulator can exercise.
 *
 * ## Why binding is done by the candidate's OWN ufrag
 *
 * A candidate belongs to whichever ICE generation the candidate itself names,
 * never to whatever epoch happens to be current when the callback fires.
 * `onIceCandidate` is asynchronous and a restart can land between gathering and
 * delivery, so labelling by a mutable "current epoch" variable attributes
 * candidates to the wrong generation — which is precisely how an old-generation
 * candidate gets mistaken for proof that a migration succeeded.
 *
 * Android's `org.webrtc.IceCandidate` exposes `sdpMid`, `sdpMLineIndex` and
 * `sdp` and NO `usernameFragment` property (verified against the
 * `io.github.webrtc-sdk:android:150.7871.01` classes). The string extension is
 * therefore the only source this platform has, which is exactly what the frozen
 * wire says the native ports use.
 */
object RelayRenewSdp {

    // ── the pin ─────────────────────────────────────────────────────────────

    /**
     * The identity a renewal must not change: the DTLS fingerprint set, the
     * m-line/mid sequence, and the answer's chosen `setup` role.
     *
     * Pinning keeps the existing DTLS peer in place. It is NOT the root of
     * trust — that remains the E2E key the SAS anchored.
     */
    data class Pin(
        /** Normalised `<hash-lower> <HEX-UPPER>`, sorted and deduplicated. */
        val fingerprints: List<String>,
        /** The `a=mid:` sequence, in order. Its SIZE is the m-line count. */
        val mids: List<String>,
        /** The `a=setup:` role, lower-cased, or "" when the description states
         *  none. */
        val setup: String,
    )

    /**
     * Read the pin out of a description.
     *
     * The fingerprint hash is lower-cased and the hex upper-cased before
     * comparison because the two ends may list the same set in a different
     * order and RFC 8122 leaves the hex case open — a strict byte comparison
     * would reject a legitimate peer for restating its own fingerprint in the
     * other case. The set is sorted and deduplicated for the same reason.
     *
     * The LAST `a=setup:` wins, matching how a media-level attribute overrides a
     * session-level one. A bundled `link/1` description states one.
     */
    fun pin(sdp: String): Pin {
        val fingerprints = LinkedHashSet<String>()
        val mids = ArrayList<String>()
        var setup = ""
        for (rawLine in sdp.split("\r\n", "\n", "\r")) {
            val line = rawLine.trim()
            when {
                line.startsWith(FINGERPRINT_PREFIX) -> {
                    val value = line.removePrefix(FINGERPRINT_PREFIX).trim()
                    val space = value.indexOf(' ')
                    if (space <= 0) continue
                    val hash = value.substring(0, space).lowercase()
                    val hex = value.substring(space + 1).trim().uppercase()
                    if (hash.isEmpty() || hex.isEmpty()) continue
                    fingerprints.add("$hash $hex")
                }
                line.startsWith(MID_PREFIX) -> mids.add(line.removePrefix(MID_PREFIX).trim())
                line.startsWith(SETUP_PREFIX) ->
                    setup = line.removePrefix(SETUP_PREFIX).trim().lowercase()
            }
        }
        return Pin(fingerprints.sorted(), mids, setup)
    }

    /**
     * Whether a renewal description may be applied against the epoch-0
     * baseline.
     *
     * `setup` is compared ONLY for an answer: an offer legitimately restates
     * `actpass`, so the role is compared where it is actually chosen. A
     * baseline that stated no role at all constrains nothing rather than
     * demanding the peer also state none.
     */
    fun pinMatches(baseline: Pin, next: Pin, isAnswer: Boolean): Boolean {
        if (baseline.fingerprints != next.fingerprints) return false
        if (baseline.mids != next.mids) return false
        if (isAnswer && baseline.setup.isNotEmpty() && baseline.setup != next.setup) return false
        return true
    }

    // ── ufrag binding ───────────────────────────────────────────────────────

    /**
     * The `a=ice-ufrag:` of a description, or "".
     *
     * A bundled `link/1` description has one. If a future description ever
     * carried two different ones this returns the FIRST, which is the
     * conservative answer: a candidate whose ufrag does not match it is
     * dropped, and dropping a usable candidate costs a failed migration that
     * keeps the old deadline — while accepting one from the wrong generation is
     * what the binding exists to stop.
     */
    fun iceUfrag(sdp: String): String {
        for (rawLine in sdp.split("\r\n", "\n", "\r")) {
            val line = rawLine.trim()
            if (line.startsWith(UFRAG_PREFIX)) return line.removePrefix(UFRAG_PREFIX).trim()
        }
        return ""
    }

    /**
     * The ufrag a CANDIDATE STRING states, or "".
     *
     * Read from the candidate's own `ufrag <x>` extension. Extensions are
     * `name value` pairs, so the scan is pairwise: a literal `ufrag` appearing
     * as a VALUE cannot be read as the key.
     */
    fun candidateUfrag(candidate: String): String {
        val parts = candidate.trim().split(WHITESPACE)
        var i = 0
        while (i + 1 < parts.size) {
            if (parts[i] == "ufrag") return parts[i + 1]
            i++
        }
        return ""
    }

    /**
     * The ufrag of an INBOUND candidate, requiring the two sources to agree.
     *
     * A candidate carries its generation twice: in the signal's
     * `usernameFragment` field, and in its own string extension. A native peer
     * may populate only the second. Requiring agreement where both exist, and
     * accepting a single source where only one does, is what makes the binding
     * work across the three clients without letting a relay relabel a candidate
     * by editing whichever copy the receiver happens to read.
     *
     * Returns "" when they contradict each other or when neither states one —
     * both of which mean the candidate cannot be attributed and must be
     * dropped.
     */
    fun inboundCandidateUfrag(candidate: String, usernameFragment: String): String {
        val embedded = candidateUfrag(candidate)
        if (embedded.isNotEmpty() && usernameFragment.isNotEmpty() && embedded != usernameFragment) {
            return ""
        }
        return if (usernameFragment.isNotEmpty()) usernameFragment else embedded
    }

    /**
     * The `typ` a candidate states — `host`, `srflx`, `prflx`, `relay` — or "".
     *
     * Used for two separate decisions, and it is worth keeping them apart:
     * whether the selected pair is RELAYED (and therefore bounded by a
     * credential at all), and whether a selected local candidate can be
     * attributed to a generation.
     */
    fun candidateType(candidate: String): String {
        val parts = candidate.trim().split(WHITESPACE)
        for (i in parts.indices) {
            if (parts[i] == "typ" && i + 1 < parts.size) return parts[i + 1]
        }
        return ""
    }

    /**
     * Whether a SELECTED LOCAL candidate demonstrably belongs to [epochUfrag].
     *
     * The candidate's own ufrag must equal this epoch's local ufrag. A selected
     * peer-reflexive candidate can carry that proof too: the SDK derives it
     * from the local port's generation. Its type alone is not a reason to
     * reject it. Missing or stale ufrags still fail closed; neither a changed
     * port nor the currently installed description substitutes for this proof.
     *
     * An empty [epochUfrag] — no local description applied yet — is never a
     * match, so observation cannot hold before there is a generation to hold
     * it against.
     */
    fun localCandidateBelongsTo(candidate: String, epochUfrag: String): Boolean {
        if (epochUfrag.isEmpty()) return false
        return candidateUfrag(candidate) == epochUfrag
    }

    /**
     * How the existing rule classifies one selected pair.
     *
     * Relay on EITHER side means the path is bounded by a TURN credential;
     * host-to-host is a LAN direct hop; anything else is a NAT-traversed direct
     * path. Identical to `classifyPath` in `web/src/lib/webrtc-core.ts`, but
     * read from the candidates the platform hands to
     * `onSelectedCandidatePairChanged` rather than scanned out of a stats
     * report — which is also why this port cannot make the mistake of picking
     * a stale nominated pair out of an unordered report.
     */
    enum class Path { RELAY, LAN, P2P }

    fun classifyPath(localCandidate: String, remoteCandidate: String): Path {
        val local = candidateType(localCandidate)
        val remote = candidateType(remoteCandidate)
        if (local == "relay" || remote == "relay") return Path.RELAY
        if (local == "host" && remote == "host") return Path.LAN
        return Path.P2P
    }

    private const val FINGERPRINT_PREFIX = "a=fingerprint:"
    private const val MID_PREFIX = "a=mid:"
    private const val SETUP_PREFIX = "a=setup:"
    private const val UFRAG_PREFIX = "a=ice-ufrag:"
    private val WHITESPACE = Regex("\\s+")
}
